lib/elixir/test/elixir/kernel/diagnostics_test.exs

```diff
@@ -750,6 +750,68 @@ defmodule Kernel.DiagnosticsTest do
       purge(Sample)
     end
 
+    @tag :tmp_dir
+    test "shows span for unused variables", %{tmp_dir: tmp_dir} do
+      path = make_relative_tmp(tmp_dir, "error_line_column.ex")
+
+      source = """
+      defmodule Sample do
+        @file "#{path}"
+
+        def foo(unused_param) do
+          :constant
+        end
+      end
+      """
+
+      File.write!(path, source)
```

Comment:
Do we need to write a file? Could we use `Code.eval_string(file: "foobar")` instead?

Reply:
we do, we currently read from files, so if the files does not exist we can't read and show the no snippet message

Reply:
Nah, ignore me, we need to do it to read the snippet.

Reply:
but I think in the future we could receive the source and work on it directly (so that we would print beautifully even in the `nofile` cases

not sure how much memory that would require though, if we would be able to do it efficiently, but I think maybe it would be faster than IO anyways

Reply:
Yeah, it is tricky. For syntax errors and others, we always have the source, but those are errors and they will be fixed anyway, so it is not worth optimizing. For the other ones, we would need a way to pass the file around throughout the compiler. It is definitely doable, perhaps using pdict, but I am unsure.

Reply:
The other thing is the warnings that come from the ParallelChecker do not have the source (they don't come from analyzing source, so those would have to read anyway).